crates/jit-icache-coherence/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+//! This crate provides utilities for instruction cache maintenance for JIT authors.
+//!
+//! In self modifying codes such as when writing a JIT, special care must be taken when marking the
+//! code as ready for execution. On fully coherent architectures (X86, S390X) the data cache (D-Cache)
+//! and the instruction cache (I-Cache) are always in sync. However this is not guaranteed for all
+//! architectures such as AArch64 where these caches are not coherent with each other.
+//!
+//! When writing new code there may be a I-cache entry for that same address which causes the
+//! processor to execute whatever was in the cache instead of the new code.
+//!
+//! See the [ARM Community - Caches and Self-Modifying Code] blog post that contains a great explanation of the above.
+//!
+//! ## Usage
+//!
+//! You should call [clear_cache] on any pages that you write with the new code that you're intending
+//! to execute. You can do this at any point in the code from the moment that you write the page up to
+//! the moment where the code is executed.
+//!
+//! You also need to call [pipeline_flush] to ensure that there isn't any invalid instruction currently
+//! in the pipeline.
+//!
+//! You can call this in a different order but you should only execute the new code after calling both.
+//!
+//! ### Example:
+//! ```
+//! # use std::ffi::c_void;
+//! # use std::io;
+//! # use jit_icache_coherence::*;
+//! #
+//! # struct Page {
+//! #   addr: *const c_void,
+//! #   len: usize,
+//! # }
+//! #
+//! # fn main() -> io::Result<()> {
+//! #
+//! # let run_code = || {};
+//! # let code = vec![0u8; 64];
+//! # let newly_written_pages = vec![Page {
+//! #    addr: &code[0] as *const u8 as *const c_void,
+//! #    len: code.len(),
+//! # }];
+//! #
+//! // Invalidate the cache for all the newly written pages where we wrote our new code.
+//! for page in newly_written_pages {
+//!     clear_cache(page.addr, page.len)?;
+//! }
+//!
+//! // Once those are invalidated we also need to flush the pipeline
+//! pipeline_flush()?;
+//!
+//! // We can now safely execute our new code.
+//! run_code();
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! <div class="example-wrap" style="display:inline-block"><pre class="compile_fail" style="white-space:normal;font:inherit;">
+//!
+//!  **Warning**: In order to correctly use this interface you *must* always call both
+//! [clear_cache] and [pipeline_flush].
+//!
+//! </pre></div>
+//!
+//! [ARM Community - Caches and Self-Modifying Code]: https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/caches-and-self-modifying-code
+
+#[cfg(all(not(target_os = "windows"), not(feature = "rustix")))]
+mod libc;
+#[cfg(all(not(target_os = "windows"), feature = "rustix"))]
+mod rustix;
+#[cfg(target_os = "windows")]
+mod win;
```

Comment:
One way I'd recommend writing this to make this more easily maintainable over time is:

```rust
cfg_if::cfg_if! {
    if #[cfg(target_os = "windows")] {
        mod win;
        use win as imp;
    } else if #[cfg(feature = "rustix")] {
        mod rustix;
        use rustix as imp;
    } else {
        mod libc;
        use libc as imp;
    }
}
```

and below just use `imp::the_method()` instead of duplicating the `#[cfg]`

Reply:
Also, how come there's a `rustix` and a `libc` implementation? Would it be reasonable to pick one as the only non-windows implementation?

Reply:
Well, for `cranelift-jit` we don't want to import rustix since that adds a bunch of dependencies for a couple of membarriers (https://github.com/bytecodealliance/wasmtime/pull/3395). For wasmtime I think we do want the safety guarantees of rustix.

So we ended up with both, one implementation in `wasmtime-jit` and another in `cranelift-jit`

Reply:
Cranelift-jit is used and tested by cg_clif.

Reply:
>Overall I assume the actual compiled-down code is basically the same modulo what function does the syscall instruction so at least form my perspective I would prefer to only have one implementation to maintain rather than two.

I hope so too! Or its probably a bug. 😄  I don't really have an opinion on what we should do here, but I'm happy with whatever.

>Also a bit more broadly I feel that the cranelift-jit crate doesn't really fit well with this repository right now. Nothing in Wasmtime uses it and it does not see heavy usage in tests I believe, but it's quite a complicated an nontrivial crate at the same time. Basically the support/maintenance story for it seems somewhat unclear but I ideally don't want it to place further burdens on other code elsewhere.

We do use the jit for all runtests in cranelift (in the filetest suite), that ends up being quite a few and it also runs on all targets that cranelift supports. And we also use the jit when fuzzing the `cranelift-fuzzgen` target.

I do think we could go the other way and try to use `cranelift-jit` in wasmtime instead of `wasmtime-jit`. That would probably be a big project, and I'm not sure if there is anything that would be fundamentally incompatible. But I think it would make sense in terms of a code sharing perspective, and as a bonus cranelift users would get a better JIT!

Reply:
I think we could share the core for handling mapping as executable and relocating, but the user facing interface of cranelift-module is designed for the C linkage model and is not compatible with the wasm linkage model.